File: backend/app/models/glucose.py
"""
Pydantic models – data schemas for Glucose API requests and responses.
"""

from typing import Optional
from pydantic import BaseModel, Field
from datetime import datetime


class GlucoseReadingResponse(BaseModel):
    timestamp: datetime
    glucose_mg_dl: float = Field(..., ge=0.0, le=600.0)
    units: str = "mg/dl"


class UploadResponse(BaseModel):
    status: str
    readings_saved: int = Field(..., ge=0)


class LatestReadingsResponse(BaseModel):
    status: str
    hours_requested: int = Field(..., ge=1, le=168)
    count: int = Field(..., ge=0)
    readings: list[GlucoseReadingResponse]


class GlucoseStats(BaseModel):
    min_glucose: float = Field(..., ge=0.0)
    max_glucose: float = Field(..., ge=0.0)
    avg_glucose: float = Field(..., ge=0.0)
    readings_count: int = Field(..., ge=0)
    time_in_range_percent: float = Field(..., ge=0.0, le=100.0)

class ClassifyRequest(BaseModel):
    glucose_mg_dl: float = Field(..., ge=0.0, le=600.0)


class ClassifyResponse(BaseModel):
    glucose_mg_dl: float
    zone: str
    is_critical: bool
    message: str

class GlucoseStatisticsResponse(BaseModel):
    count: int = Field(..., ge=0)
    min_glucose: Optional[float] = None
    max_glucose: Optional[float] = None
    avg_glucose: Optional[float] = None
    std_dev: Optional[float] = None
    time_in_range_percent: float = Field(..., ge=0.0, le=100.0)

File: frontend/src/types/glucose.ts
export interface GlucoseReading {
  timestamp: string
  glucose_mg_dl: number
  units: string
}

export interface UploadResponse {
  status: string
  readings_saved: number
  min_glucose: number | null
  max_glucose: number | null
  avg_glucose: number | null
  std_dev: number | null
  time_in_range_percent: number | null
  gmi: number | null
  cv_percent: number | null
  cv_is_stable: boolean | null
}

export interface LatestReadingsResponse {
  status: string
  hours_requested: number
  count: number
  readings: GlucoseReading[]
}

export interface GlucoseStatisticsResponse {
  count: number
  min_glucose: number | null
  max_glucose: number | null
  avg_glucose: number | null
  std_dev: number | null
  time_in_range_percent: number
  gmi: number | null
  cv_percent: number | null
  cv_is_stable: boolean | null
}

export interface ClassifyResponse {
  glucose_mg_dl: number
  zone: string
  is_critical: boolean
  message: string
}

export interface ComputedStats {
  min_glucose: number
  max_glucose: number
  avg_glucose: number
  std_dev: number | null
  time_in_range_percent: number
  gmi: number | null
  cv_percent: number | null
  cv_is_stable: boolean | null
  count: number
}

export interface PredictionPoint {
  minutes_ahead: number
  glucose_mg_dl: number
  lower_mg_dl: number
  upper_mg_dl: number
}

export interface PredictResponse {
  status: string
  subject_id: string
  horizon_steps: number
  predictions: PredictionPoint[]
  last_known_glucose: number
  model_version: string
}